odoo/addons/base/models/res_users.py

```diff
@@ -1618,6 +1619,27 @@ def remove(self):
             return {'type': 'ir.actions.act_window_close'}
         raise AccessError(_("You can not remove API keys unless they're yours or you are a system user"))
 
+    def get_api_key(self, key):
+        index = key[:INDEX_SIZE]
+        self.env.cr.execute('SELECT id, key FROM res_users_apikeys WHERE index = %s', [index])
+        for [id, current_key] in self.env.cr.fetchall():
+            if KEY_CRYPT_CONTEXT.verify(key, current_key):
+                return self.browse(id)
+
+    @api.model
+    def _generate(self, name, scope=None):
+        # no need to clear the LRU when *adding* a key, only when removing
+        k = binascii.hexlify(os.urandom(API_KEY_SIZE)).decode()
+        self.env.cr.execute("""
+        INSERT INTO res_users_apikeys (name, user_id, scope, key, index)
+        VALUES (%s, %s, %s, %s, %s)
+        RETURNING id
+        """, [name, self.env.user.id, scope, hash_api_key(k), k[:INDEX_SIZE]])
+        record_id = self.env.cr.fetchone()
+        self.env['res.users.apikeys'].invalidate_cache(ids=record_id)
```

Comment:
This weirds me out. I know that it matches the original but it doesn't seem useful, since we're creating the record using raw SQL it has no reason to be in the cache. I probably made a mistake in the original. but not quite sure what the intent was.